discovery.go

```diff
@@ -0,0 +1,225 @@
+	"github.com/libp2p/go-libp2p-core/peer"
+	"sync"
+	"time"
+)
+
+// discover represents the discovery pipeline.
+// The discovery pipeline handles advertising and discovery of peers
+type discover struct {
+	p *PubSub
+
+	// discovery assists in discovering and advertising peers for a topic
+	discovery discovery.Discovery
+
+	// advertising tracks which topics are being advertised
+	advertising map[string]context.CancelFunc
+
+	// bootstrapped tracks which topics have been bootstrapped
+	bootstrapped sync.Map
+
+	// discover handles continuing peer discovery
+	discover chan *Discover
+
+	// ongoingDiscovery tracks ongoing discovery requests
+	ongoingDiscovery map[string]struct{}
+
+	// doneDiscovery handles completion of a discovery request
+	doneDiscovery chan string
+}
+
+// newValidation creates a new validation pipeline
+func newDiscover() *discover {
+	return &discover{
+		bootstrapped:     sync.Map{},
+		discover:         make(chan *Discover),
+		ongoingDiscovery: make(map[string]struct{}),
+		doneDiscovery:    make(chan string),
+		advertising:      make(map[string]context.CancelFunc),
+	}
+}
+
+// Start attaches the discovery pipeline to a pubsub instance and starts event loop
+func (d *discover) Start(p *PubSub) {
+	d.p = p
+	go d.discoverLoop()
+}
+
+func (d *discover) discoverLoop() {
+	for {
+		select {
+		case discover := <-d.discover:
+			topic := discover.topic
+			if _, ok := d.ongoingDiscovery[topic]; !ok {
+				d.ongoingDiscovery[topic] = struct{}{}
+				go func() {
+					d.handleDiscovery(d.p.ctx, topic, discover.opts)
+					d.doneDiscovery <- topic
+				}()
+			}
+		case topic := <-d.doneDiscovery:
+			delete(d.ongoingDiscovery, topic)
+		case <-d.p.ctx.Done():
+			log.Info("pubsub discoverloop shutting down")
+			return
+		}
+	}
+}
+
+// advertise advertises this node's interest in a topic to a discovery service
+func (d *discover) advertise(topic string) {
+	if d.discovery != nil {
+		advertisingCtx, cancel := context.WithCancel(d.p.ctx)
+
+		if _, ok := d.advertising[topic]; ok {
+			return
+		}
+		d.advertising[topic] = cancel
+
+		go func() {
+			next, err := d.discovery.Advertise(advertisingCtx, topic)
+			if err != nil {
+				log.Warningf("bootstrap: error providing rendezvous for %s: %s", topic, err.Error())
+			}
+
+			t := time.NewTimer(next)
+			for {
+				select {
+				case <-t.C:
+					next, err = d.discovery.Advertise(advertisingCtx, topic)
+					if err != nil {
+						log.Warningf("bootstrap: error providing rendezvous for %s: %s", topic, err.Error())
+					}
+					t.Reset(next)
+				case <-advertisingCtx.Done():
+					t.Stop()
+					return
+				}
+			}
+		}()
+	}
+}
+
+func (d *discover) stopAdvertise(topic string) {
+	if advertiseCancel, ok := d.advertising[topic]; ok {
+		advertiseCancel()
+		delete(d.advertising, topic)
+	}
+}
+
+// bootstrap handles initial bootstrapping of peers for a given topic
+func (d *discover) bootstrap(async bool, topic string, dOpts ...discovery.Option) {
+	newMux := &sync.Mutex{}
```

Comment:
I don't understand the use of this mutex; it doesn't seem to be protecting any critical section as it's a brand new object each time.

Reply:
It ensures that bootstrapping only occurs once per topic. `sync.Map.LoadOrStore` ensures that an item will only be added to the map once. This adding to the map indicates that bootstrapping has started, but we have no way to indicate it has completed.

The mutex does this by being Locked until bootstrapping completes at which point it is released. We could do this in other ways such as by having a second sync.Map also that indicates "bootstrapping has completed".

If you have a preference I'm open to suggestions.

Reply:
but you create a new mutex every time you call bootstrap, there is nothing protected by this.

Reply:
clarification: the map signifies "have we started bootstrapping" and the mutex controls making sure we're only going to bootstrap once,

After discussing with @vyzo we're going to see how this looks if we use maps of (closed) channels instead of maps of mutexes.